src/search.jl

```diff
@@ -1,81 +1,119 @@
 
-end
+# end
 
-function _merge(
-     model::Model,
-     sol::Solution,
-    )
-    boundary = zeros(Int, length(sol.states))
-    for (i, σ) ∈ enumerate(sol.states)
-        boundary[i] = generate_boundary(model, σ)
-    end
+# function _merge(
+#      network::AbstractGibbsNetwork,
+#      sol::Solution,
+#     )
+#     boundary = []
+#     for (i, σ) ∈ enumerate(sol.states)
+#         push!(boundary, generate_boundary(network, σ))
+#     end
 
-    idx = _partition_into_unique(boundary, sol.energies)
-    Solution(sol.energies[idx], sol.states[idx], sol.probabilities[idx])
-end
+#     idx = _partition_into_unique(boundary, sol.energies)
+#     Solution(
+#         sol.energies[idx],
+#         sol.states[idx],
+#         sol.probabilities[idx],
+#         sol.largest_discarded_probability)
+# end
 
-function _sort(
-    sol::Solution,
-    k::Int,
+#TODO: this can probably be done better
+function _branch_state(
+    cfg::Vector,
+    state::Vector,
+    basis::Vector,
     )
-    perm = partialsortperm(sol.probabilities, 1:k, rev=true)
-    Solution(sol.energies[perm], sol.states[perm], sol.probabilities[perm])
+    tmp = Vector{Int}[]
+    for σ ∈ basis push!(tmp, vcat(state, σ)) end
+    vcat(cfg, tmp)
 end
 
-function _update_solution!(
-    sol::Solution,
-    model::Model, 
-    )
-    prob = eng = psol = [] 
+@inline _init_solution() = (Float64[], Float64[], Vector{Int}[])
 
-    for (i, σ) ∈ enumerate(sol.states)
-        p = conditional_probability(model, σ)
-        push!(prob, sol.probabilities[i] .* p)
-        push!(eng, sol.energies[i] .+ δE(model, σ))
-        # add sol
-    end
+# TODO: logic here can probably be done better
+function _bound(pdo::Vector{Float64}, cut::Int)
+    k = length(pdo)
+    second_phase = false
+    if k > cut + 1 k = cut + 1; second_phase = true end
 
-    sol = Solution(vec(eng), sol, vec(prob)) 
-    _sort(_merge(model, sol), k)
+    idx = partialsortperm(pdo, 1:k, rev=true)
+
+    if second_phase
+        return idx[1:end-1], pdo[last(idx)]
+    else
+        return idx, -Inf
+    end
 end
 
-@inline function _largest_discarded_probability!(
-    lpCut::Float64,
+function _branch_and_bound(
     sol::Solution,
+    network::AbstractGibbsNetwork,
+    node::Int,
+    cut::Int,
     )
-    p = last(sol.probabilities)
-    lpCut < p ? lpCut = p : ()
+
+    # branch
+    pdo, eng, cfg = _init_solution()
+    k = get_prop(network.fg, node, :loc_dim)
+
+    # for (p, σ, e) ∈ zip(sol.probabilities, sol.states, sol.energies)
+    #     pdo = [pdo; p .* conditional_probability(network, σ)]
+    #     eng = [eng; e .+ update_energy(network, σ)]
+    #     cfg = _branch_state(cfg, σ, collect(1:k))
+    #  end
+
+    for (p, σ, e) ∈ zip(sol.probabilities, sol.states, sol.energies)
+        pdo = [pdo; p .* conditional_probability(network, σ)]
+        eng = [eng; e .+ update_energy(network, σ)]
+        cfg = _branch_state(cfg, σ, collect(1:k))
+     end
+
+    # bound
+    K, lp = _bound(pdo, cut)
```

Comment:
Given that `K` are indices, and `lp` is (from what I understand) lowest probability, can we just rename them to `indices` and `lowest_prob` or something similar to enhance readability?

Reply:
Done.